### DISCUSSION TITLE
ЛР 4 для ИВТ. Вопросы по реализации памяти

### DISCUSSION THREAD

**Discussion by mrz25** [MPSU/exAPS#31] [Вопросы и ответы] [answered]
Здравствуйте, у меня встал вопрос по поводу 4 лабы, читал я ее читал и не врубился по поводу того, что у нас с памятью и ее адресами происходит, то есть раньше у нас на Instruction Memory поступал 8-битный адрес и мы по этому адресу обращались к байту в память, потому что память у нас тогда была [8:0] mem [255:0] и затем приплюсовывали к этому байту следующие три байта, чтобы получить полное слово
а теперь мы подаем туда 32-битный адрес, зачем-то сдвигаем его на 2 бита вправо и как бы да, а что дальше? как у нас память щас выглядеть должна  "[8:0] mem [255:0]" или сколько, или она у нас уже пословная становится, а не побайтовая? Или у нас подается 32-бита побайтовой адресации, а внутри у нас уже пословная [31:0] mem [31:0]?
![image](https://user-images.githubusercontent.com/71031775/196678746-8edf0860-ef3b-4190-a37d-edec2de5d31a.png)
![image](https://user-images.githubusercontent.com/71031775/196678933-602c938f-5ab7-44a9-ad9a-05dc26729b2a.png)

**Comment by HepoH3** [comment, marked as answer]
Тут коллизия произошла, эта методичка предполагает, что вы раньше делали память пословную (`reg [31:0] instr_mem [0:255]`. И чтобы такая память заработала в архитектуре RISC-V, необходимо выполнить пару хаков (о которых написано в методичке). На лабах, я вам сразу предлагал делать побайтовую память инструкций, чтобы эти трюки не выполнять, так что можете их игнорить (речь идет только о хаках памяти инструкций).

**Reply by mrz25** [reply]
Ага, то есть в итоге (как я понял) у нас и на Instruction Memory и на Data Memory на input A подаётся 8-битный адрес? Как и было раньше

**Reply by HepoH3** [reply]
Не, эта часть остается. Тут фишка в том, что вообще, адреса памяти в RV32I 32-битные (диапазон памяти от 0 до 4Гб), но разумеется вы не сможете сделать такую память на регистрах, поэтому вам предлагалось сделать память на 256 слов (для которых надо только 8 бит адреса). Кстати, если вы делали память шириной 8 бит, то число ячеек сейчас надо сделать 1024 (в 1024 байтах хранится 256 слов).
И вот вам предлагают увеличить разрядность адреса, не менять внутри памяти количество ячеек, но делать вид будто ячеек этих больше. Причем не просто делать вид, а прицепить ваши ячейки с старшим адресам, а по всем остальным адресам возвращать 0.
Т.е. если вам подается адрес `NN00000` — вы должны вернуть нулевое слово, если `NN00001` — первое, `NN003FC` — последнее. А для всех остальных адресов вернуть 0. Сделать это очень легко, ведь по сути адреса ячеек почти те же самые, только к ним приклеили слева `NN00`.

**Reply by gr33nka** [reply]
@mrz25 можешь глянуть в методичку второй лабы,  посмотреть что там было сделано и все встанет на свои места

**Reply by mrz25** [reply]
То есть адрес у нас теперь становится 32-битным (потому что стандарт такой), при этом внутри файла мы его сдвигаем на 2 бита, чтобы получить 8-битный адрес и обратиться к первому байту слова из [7:0] mem [1023:0], а затем так же внутри файла делаем конкатенацию с последующими тремя байтами, но тогда почему мы их начинаем называть как 0, 1, 2, 3... Или это все очень условно и лишь для того, чтобы сказать, мол, адресация у нас пословная теперь

**Reply by HepoH3** [reply]
Не, забудь про сдвиг, это нужно было бы, если ты память была `reg [31:0] mem [0:255]` вместо `reg [7:0] mem [0:1023]`

**Reply by HepoH3** [reply]
Адрес 32-битный из-за спеки, да. Конкатенация байтов, потому что инструкция тоже 32-битная, а одна ячейка памяти `reg [7:0] mem [0:1023]` занимает один байт

**Reply by HepoH3** [reply]
И поскольку 32-битным адресом можно адресовать 4млрд байт, а в памяти `reg [7:0] mem [0:1023]` всего один кибибайт, мы делаем хак, что этот кибибайт располагается только по адресам `NN00000:NN003FC`, а на всех остальных адресах просто возвращаем ноль.

**Reply by gr33nka** [reply]
Лишние биты адреса не участвуют в формировании адреса памяти `reg [7:0] mem [0:1023]`. Эти откинутые биты участвуют в проверке того, что адреса в диапазоне `NN00000:NN003FC`

**Reply by mrz25** [reply]
то есть грубо говоря у нас диапазон адресов от 000 в шестнадцатиричной до 3FC но при этом 3FC как бы выходит за диапазон значащих 8 бит и занимает 12 бит, в которые мы никак не уложимся при задании максимального 8-битного адреса - 11111111 (FF) и как тогда обращаться к памяти по адресу 3FB (11 1111 1011), беря [7:0] бит адреса 32-битного, если оказывается он будет неправильно воспринят как FB (1111 1011)?

**Reply by HepoH3** [reply]
> то есть грубо говоря у нас диапазон адресов от 000 в шестнадцатиричной до 3FC но при этом 3FC как бы выходит за диапазон значащих 8 бит и занимает 12 бит, в которые мы никак не уложимся при задании максимального 8-битного адреса - 11111111 (FF) и как тогда обращаться к памяти по адресу 3FB (11 1111 1011), беря [7:0] бит адреса 32-битного, если оказывается он будет неправильно воспринят как FB (1111 1011)?

3FC занимает 10 бит (первая шестнадцатиричная цифра содержит всего две младшие единицы, старшие нули как бы отбрасываются (`0x3FC == 0b11_1111_1100 <- 10 бит`). При этом, данный адрес побайтовый, а память предполагается пословной. Поэтому, все приходящие адреса будут иметь на конце `00`, которые тоже можно отбросить. Таким образом получаем эффективное использование ровно восьми бит.

**Reply by HepoH3** [reply]
![изображение](https://user-images.githubusercontent.com/17159587/198833179-b192ed73-efaa-4c18-9a66-df6851558847.png)

P.S. на выход идут данные с 255ой ячейки Memory, но проц думает, что получает данные слова, нулевой байт которого размещен по адресу `0xNN003fc`
